drivers/dma/dma_silabs_ldma.c

```diff
@@ -0,0 +1,564 @@
+		    config->block_count == 1) {
+			LOG_WRN("DMA scatter_gather enabled but there is only one descriptor "
+				"configured in this channel(%u)",
+				chan_conf->id);
+		}
+
+		desc->xfer.structType = ldmaCtrlStructTypeXfer;
+		desc->xfer.structReq = 1;
+		ret = dma_silabs_get_transfer_parameter(
+			config->source_data_size, config->dest_data_size, block->block_size, desc);
+		if (ret) {
+			return ret;
+		}
+
+		desc->xfer.byteSwap = 0;
+
+		/* Warning : High LDMA blockSize (high burst) mean a large transfer
+		 *           without LDMA controller re-arbitration.
+		 */
+		ret = dma_silabs_get_blocksize(config->source_burst_length,
+					       config->dest_burst_length, config->source_data_size,
+					       desc);
+		if (ret) {
+			return ret;
+		}
+
+		/* if complete_callbacks_enabled, callback is called at then end of each descriptors
+		 * list (block for zephyr)
+		 */
+		desc->xfer.doneIfs = config->complete_callback_en ? 1 : 0;
+		desc->xfer.reqMode = ldmaCtrlReqModeAll;
+		desc->xfer.decLoopCnt = 0;
+		desc->xfer.ignoreSrec = block->flow_control_mode;
+
+		/* In silabs LDMA, increment sign is managed with the transfer configuration
+		 * which is common for all descs of the channel. Zephyr DMA API allows
+		 * to manage increment sign for each block desc which can't be done with
+		 * silabs LDMA. If increment sign is different in 2 block desc, then an
+		 * error is returned.
+		 */
+		if (block->source_addr_adj != DMA_ADDR_ADJ_NO_CHANGE &&
+		    block->source_addr_adj != chan_conf->xfer_config.ldmaCfgSrcIncSign) {
+			return -ENOTSUP;
+		}
+
+		if (block->source_addr_adj == DMA_ADDR_ADJ_NO_CHANGE) {
+			desc->xfer.srcInc = ldmaCtrlSrcIncNone;
+		} else {
+			desc->xfer.srcInc = ldmaCtrlSrcIncOne;
+		}
+		if (block->dest_addr_adj == DMA_ADDR_ADJ_NO_CHANGE) {
+			desc->xfer.dstInc = ldmaCtrlDstIncNone;
+		} else {
+			desc->xfer.dstInc = ldmaCtrlDstIncOne;
+		}
+
+		desc->xfer.srcAddrMode = ldmaCtrlSrcAddrModeAbs;
+		desc->xfer.dstAddrMode = ldmaCtrlDstAddrModeAbs;
+
+		if (config->head_block->source_address == 0) {
+			LOG_WRN("source_buffer address is null.");
+		}
+
+		if (config->head_block->dest_address == 0) {
+			LOG_WRN("dest_buffer address is null.");
+		}
+
+		desc->xfer.srcAddr = block->source_address;
+		desc->xfer.dstAddr = block->dest_address;
+
+		if (block->next_block) {
+			desc->xfer.linkMode = ldmaLinkModeAbs;
+			desc->xfer.link = 1;
+
+			ret = sys_mem_blocks_alloc(data->dma_desc_pool, 1, (void **)&temp_desc);
+			if (ret) {
+				return -EXFULL;
+			}
+
+			/* Link to the next descriptor and configure it in the next iteration */
+			desc->xfer.linkAddr = LDMA_DESCRIPTOR_LINKABS_ADDR_TO_LINKADDR(temp_desc);
+			desc = temp_desc;
+			block = block->next_block;
+			/* Protection against descriptor loop*/
+			if (block == head_block) {
+				block = 0;
+			}
+		} else {
+			desc->xfer.linkMode = 0;
+			desc->xfer.link = 0;
+			desc->xfer.linkAddr = 0;
+			block = 0;
+		}
+	} while (block);
```

Comment:
This loop is rather long. I feel you could make the yor code simpler with a function like:
```
static LDMA_Descriptor_t *dma_silabs_get_descriptor(struct dma_config *config, struct dma_block_config *block);
```